File: src/main/java/io/github/willslaunwhite/linkedlist/problems/MergeTwoSortedLists.java
package io.github.willslaunwhite.linkedlist.problems;

import io.github.willslaunwhite.linkedlist.ListNode;

/*
 * You are given the heads of two sorted linked lists list1 and list2.
 * Merge the two lists in a one sorted list. The list should be made by splicing together the nodes of the first two lists.
 * Return the head of the merged linked list. 
 * 
 * Input: list1 = [1,2,4], list2 = [1,3,4]
 * Output: [1,1,2,3,4,4] 
 * 
 * Input: list1 = [], list2 = []
 * Output: []
 * 
 * Input: list1 = [], list2 = [0]
 * Output: [0]
 */

public class MergeTwoSortedLists {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        final ListNode head = new ListNode(0);
        ListNode prev = head;

        while (list1 != null && list2 != null) {
            if (list1.val < list2.val) {
                prev.next = list1;
                list1 = list1.next;
            } else {
                prev.next = list2;
                list2 = list2.next;
            }
            prev = prev.next;
        }
        prev.next = list1 != null ? list1 : list2;
        return head.next;
    }

    public ListNode mergeTwoListsRecursive(ListNode list1, ListNode list2) {
        if (list1 == null) {
            return list2;
        }
        if (list2 == null) {
            return list1;
        }

        if (list1.val < list2.val) {
            list1.next = mergeTwoListsRecursive(list1.next, list2);
            return list1;
        } else {
            list2.next = mergeTwoListsRecursive(list1, list2.next);
            return list2;
        }
    }
}

File: python/linked_list/problems/merge_two_sorted_lists.py
from typing import Any, Optional
from linked_list.list_node import ListNode


def merge_two_sorted_lists_iterative(
    list1: Optional[ListNode], list2: Optional[ListNode]
) -> Optional[ListNode]:
    """
    Given the heads of two sorted linked lists, return the head of
    a single sorted list

    Args:
        list1 (Optional[ListNode]): head of first list
        list2 (Optional[ListNode]): head of second list

    Returns:
        head of a single sorted list
    """
    head: Optional[ListNode] = ListNode(0)
    prev: Optional[ListNode] = head

    while list1 and list2:
        if list1.val < list2.val:
            prev.next = list1
            list1 = list1.next
        else:
            prev.next = list2
            list2 = list2.next
        prev = prev.next
    prev.next = list1 or list2

    return head.next


def merge_two_sorted_lists_recursive(
    list1: Optional[ListNode], list2: Optional[ListNode]
) -> Optional[ListNode]:
    if not list1 or not list2:
        return list1 or list2

    if list1.val < list2.val:
        list1.next = merge_two_sorted_lists_recursive(list1.next, list2)
        return list1
    else:
        list2.next = merge_two_sorted_lists_recursive(list1, list2.next)
        return list2
